rfd/0135-kube-apps-discovery.md

```diff
@@ -0,0 +1,227 @@
+    state: draft
+---
+
+# RFD 135 - Kubernetes App discovery
+
+## Required Approvers
+
+- Engineering: `@r0mant` && `@tigrato`
+- Product: `@klizhentas` || `@xinding33`
+- Security: `@reedloden` || `@jentfoo`
+
+## What
+
+Proposes discovery of apps that are hosted inside Kubernetes clusters and exposed as services.
+
+Related issue: [#25538](https://github.com/gravitational/teleport/issues/25538)
+
+## Why
+
+Some users dynamically deploy apps into their Kubernetes clusters and expose those as Kubernetes services. When they actively
+create/update/remove apps, manually trying to mirror these changes to Teleport may be time-consuming and prone to
+errors, especially for organizations that have a large number of services running in their Kubernetes clusters.
+Teleport will be able to discover Kubernetes service and automatically register them as
+Teleport apps and keep them in up to date state, simplifying process and making user experience much better.
+
+## Scope
+
+In this document we focus on dynamically creating Teleport apps based on discovery of services in Kubernetes clusters that are running
+Teleport agents. Discovery service and app service should be running inside the Kubernetes cluster whose services are being exposed.
+
+## Details
+
+Kubernetes app discovery will build upon existing infrastructure for other discovery capabilities that are already there. We will add another
+type of matchers to the discovery service configuration - `kubernetes`. If these matchers are present, every 5 minutes discovery service will 
+poll Kubernetes cluster, inside which it is running, for a list of services. Based on this list, it will form a list of Teleport apps resources
+and update it on the backend, constantly keeping the dynamic apps state up to date. Writing to the backend will be handled
+by already existing mechanism in our code - reconciler. Reconciler will compare given fresh list of resources and resources we currently have
+on the backend, and then will decide on appropriate actions (create/update/delete) that are required on
+the backend to change resources to the desired state.
+The Teleport app service that runs in the same Kubernetes cluster will then react to those changes, and will start/stop proxying
+apps that are selected by appropriate labels.
+
+Name of the created Teleport app will consist of Kubernetes service name, namespace and cluster name:
+`$SERVICE_NAME-$CLUSTER_NAME-$NAMESPACE`. This is done to avoid naming collisions in discovered apps. Though app names
+will be long in that case, there's ongoing work to improve UX when working with long resources names in Teleport, see [RFD 129](https://github.com/gravitational/teleport/pull/27258).
+
+We will add capability to use optional annotations to better control transformation of Kubernetes services into Teleport apps.
+By default we will be exposing services as `tcp` apps, since it's the most general type, but annotation or exposed port's `appProtocol` field
```

Comment:
Can we add well-known HTTP ports to this detection mechanism as well and automatically create HTTP apps from them without requiring an annotation? 80, 443, 8080, etc.

Reply:
Yep, I've added it to the description of `teleport.dev/discovery-protocol` annotation 👍 (will include port 8080 as well though)